File: julia/complex-numbers/complex-numbers.jl
struct ComplexNumber <: Number
    rp::Float64
    ip::Float64
end

ComplexNumber(rp) = ComplexNumber(rp, 0)
const jm = ComplexNumber(0, 1)

real(c) = c.rp
imag(c) = c.ip

Base.:+(c1::ComplexNumber, c2::ComplexNumber) = ComplexNumber(real(c1) + real(c2), imag(c1) + imag(c2))
Base.:+(c::ComplexNumber, n::Real) = c + ComplexNumber(n)
Base.:+(n::Real, c::ComplexNumber) = c + n
Base.:-(c1::ComplexNumber, c2::ComplexNumber) = ComplexNumber(real(c1) - real(c2), imag(c1) - imag(c2))
Base.:*(c1::ComplexNumber, c2::ComplexNumber) = ComplexNumber(real(c1) * real(c2) - imag(c1) * imag(c2), real(c1) * imag(c2) + imag(c1) * real(c2))
Base.:*(c::ComplexNumber, n::Real) = c * ComplexNumber(n)
Base.:*(n::Real, c::ComplexNumber) = ComplexNumber(n) * c
Base.:/(c1::ComplexNumber, c2::ComplexNumber) = (de = real(c2)^2 + imag(c2)^2; ComplexNumber((real(c1) * real(c2) + imag(c1) * imag(c2)) / de, (imag(c1) * real(c2) - real(c1) * imag(c2)) / de))
Base.:^(c::ComplexNumber, p::Int) = reduce(*, repeat([c], mod(p, 4)))
Base.:≈(c1::ComplexNumber, c2::ComplexNumber) = isapprox(real(c1), real(c2), atol=eps() * 2) && isapprox(imag(c1), imag(c2), atol=eps() * 2)

Base.abs(c::ComplexNumber) = √(real(c)^2 + imag(c)^2)
Base.conj(c::ComplexNumber) = ComplexNumber(real(c), (i = imag(c); i == 0 ? 0 : -i))
Base.exp(c::ComplexNumber) = ComplexNumber(ℯ^real(c)) * ComplexNumber(cos(imag(c)), sin(imag(c)))

File: csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double _real;
    private readonly double _imag;

    public ComplexNumber(double real, double imaginary = 0) => (_real, _imag) = (real, imaginary);

    public double Real() => _real;
    public double Imaginary() => _imag;

    public ComplexNumber Mul(ComplexNumber other) => new(_real * other._real - _imag * other._imag, _real * other._imag + other._real * _imag);
    public ComplexNumber Add(ComplexNumber other) => new(_real + other._real, _imag + other._imag);
    public ComplexNumber Sub(ComplexNumber other) => new(_real - other._real, _imag - other._imag);
    public ComplexNumber Div(ComplexNumber other)
    {
        var otherPow = Math.Pow(other._real, 2) + Math.Pow(other._imag, 2);
        return new((_real * other._real + _imag * other._imag) / otherPow, (_imag * other._real - _real * other._imag) / otherPow);
    }

    public double Abs() => Math.Sqrt(_real * _real + _imag * _imag);
    public ComplexNumber Conjugate() => new(_real, -_imag);
    public ComplexNumber Exp() => new ComplexNumber(Math.Pow(Math.Exp(1), _real)).Mul(new ComplexNumber(Math.Cos(_imag), Math.Sin(_imag)));
}